deepchem/models/tensorgraph/layers.py

```diff
@@ -0,0 +1,239 @@
+    parent_tensor = parent.out_tensor
+    self.out_tensor = tf.reshape(parent_tensor, shape=(-1, vector_size))
+    return self.out_tensor
+
+
+class Reshape(Layer):
+
+  def __init__(self, shape, **kwargs):
+    self.shape = shape
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    parent_tensor = parents[0].out_tensor
+    self.out_tensor = tf.reshape(parent_tensor, self.shape)
+
+
+class CombineMeanStd(Layer):
+
+  def __init__(self, **kwargs):
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    if len(parents) != 2:
+      raise ValueError("Must have two parents")
+    mean_parent, std_parent = parents[0], parents[1]
+    mean_parent_tensor, std_parent_tensor = mean_parent.out_tensor, std_parent.out_tensor
+    sample_noise = tf.random_normal(
+        mean_parent_tensor.get_shape(), 0, 1, dtype=tf.float32)
+    self.out_tensor = mean_parent_tensor + (std_parent_tensor * sample_noise)
+
+
+class Repeat(Layer):
+
+  def __init__(self, n_times, **kwargs):
+    self.n_times = n_times
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    if len(parents) != 1:
+      raise ValueError("Must have one parent")
+    parent_tensor = parents[0].out_tensor
+    t = tf.expand_dims(parent_tensor, 1)
+    pattern = tf.stack([1, self.n_times, 1])
+    self.out_tensor = tf.tile(t, pattern)
+
+
+class GRU(Layer):
+
+  def __init__(self, n_hidden, out_channels, batch_size, **kwargs):
+    self.n_hidden = n_hidden
+    self.out_channels = out_channels
+    self.batch_size = batch_size
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    if len(parents) != 1:
+      raise ValueError("Must have one parent")
+    parent_tensor = parents[0].out_tensor
+    gru_cell = tf.nn.rnn_cell.GRUCell(self.n_hidden)
+    initial_gru_state = gru_cell.zero_state(self.batch_size, tf.float32)
+    rnn_outputs, rnn_states = tf.nn.dynamic_rnn(
+        gru_cell,
+        parent_tensor,
+        initial_state=initial_gru_state,
+        scope=self.name)
+    projection = lambda x: tf.contrib.layers.linear(x, num_outputs=self.out_channels, activation_fn=tf.nn.sigmoid)
+    self.out_tensor = tf.map_fn(projection, rnn_outputs)
+
+
+class TimeSeriesDense(Layer):
+
+  def __init__(self, out_channels, **kwargs):
+    self.out_channels = out_channels
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    if len(parents) != 1:
+      raise ValueError("Must have one parent")
+    parent_tensor = parents[0].out_tensor
+    dense_fn = lambda x: tf.contrib.layers.fully_connected(x, num_outputs=self.out_channels,
+                                                           activation_fn=tf.nn.sigmoid)
+    self.out_tensor = tf.map_fn(dense_fn, parent_tensor)
+
+
+class Input(Layer):
+
+  def __init__(self, shape, **kwargs):
+    self.t_shape = shape
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    self.out_tensor = tf.placeholder(tf.float32, shape=self.t_shape)
```

Comment:
Since these are new classes, you might want to design them from the beginning to be able to work with queues.  That means not hardcoding placeholders like this.  During evaluation, the inputs to the graph are indeed placeholders.  But during training, the placeholders are the inputs to the queue, and the outputs from the queue become the inputs to the graph.  So it needs to be able to do either one, depending on which purpose you're building the graph for.  See #488.